teps/0087-custom-tasks-graduation.md

```diff
@@ -0,0 +1,309 @@
+Tekton Pipelines Owners will be the Owners of this repository, and each `Custom Task` will have its own Owners. 
+
+The [*beta* API policy][api-policy] applies to the *stable* `Custom Tasks`, meaning that:
+- Any [backwards incompatible][backwards] changes must be introduced in a backwards compatible manner first, with a 
+deprecation warning in the release notes and migration instructions.
+- Users will be given at least 9 months to migrate before a backward incompatible change is made.
+- Backwards incompatible changes have to be approved by more than half of the `Custom Task`'s Owners.
+
+The *Tekton Custom Tasks* repository will provide the testing infrastructure needed by stable *Custom Tasks*. We will make 
+monthly releases of `Custom Tasks` that have had changes since their latest release, which we can automate. Moreover, 
+we will have nightly releases of each `Custom Task` to catch any failures. Each *stable* `Custom Task` will have 
+detailed documentation and multiple examples, and be used in dogfooding if we have use cases for it.
+
+We can explore a bundling releases of *Tekton Pipelines* and its latest compatible *stable* `Custom Tasks`, such that 
+users can install *Tekton Pipelines* and `Custom Tasks` together.
+
+#### Graduation Requirements
+
+We can consider graduating a given *experimental* `Custom Task` to *stable* if it meets the following requirements:
+
+1. It is thoroughly tested.
+2. It has nightly releases.
+3. It provides several examples or samples.
+4. It has detailed documentation for installation and usage.
+5. It is used in dogfooding if we have use cases for it.
+6. It has received feedback from the community (solicited in working group, mailing list or other channels).
+
+#### Graduation Process
+
+The graduation process from *experimental* to *stable* for a given `Custom Task` would be:
+
+1. Its Owners validate that the *experimental* `Custom Task` meets the above requirements.
+2. Its Owners open a Tekton Enhancement Proposal (TEP) for the functionality provided by the `Custom Task` with:
+   1. A nomination for graduation from *experimental* to *stable*.
+   2. The motivation (goals, use cases and requirements).
+   3. The proposal (syntax, design details, design evaluation and alternatives).
+3. Its Owners nominate the `Custom Task` for graduation from *experimental* to *stable*.
+4. When the *Tekton Pipelines* Owners have approved the TEP as implementable, the `Custom Task`'s Owners will:
+   1. Migrate the `Custom Task` from the *Tekton Experimental* repository to the *Tekton Custom Tasks* repository.
+   2. Mark the *experimental* to *stable* TEP as *implemented* 
+
+### Integrated
+
+When a given *stable* `Custom Task` is actively used and needed for common Continuous Delivery use cases, we can 
+consider making it *integrated* - meaning that its functionality is natively supported in the *Tekton Pipelines API*. 
```

Comment:
This requirement is not 100% clear to me, I have a few questions.

Does this mean adding some syntactic sugar in the API so that it becomes easier for users to use? 

Do integrated custom tasks continue to rely on the `Run` CRD? If so, is the intention of the API changes to make that transparent to users?

In terms of code, does the custom task continue to live in a dedicated repo? Does the ownership move to the entire pipeline team? Is the custom task released together with pipeline ?

What kind of backward compatibility we provide. If a custom task becomes integrated, do users have to change their pipelines in order to benefit from the latest versions?

Finally, what level of stability is required for a custom task to have tasks and pipelines that depend on it in our catalog?

Reply:
> This requirement is not 100% clear to me, I have a few questions.
> 

Sure, happy to answer them and clarify in the TEP :) 

> 1. Does this mean adding some syntactic sugar in the API so that it becomes easier for users to use?
> 

It's more than that, it means that the functionality that the `Custom Task` provides is integrated into  *Tekton Pipelines*, that is, the functionality is directly / natively supported in the *Tekton Pipelines API* without the `Custom Task` / `Run`

Taking the example of `Pipelines` in `Pipelines`:
- In *stable*, users rely on `Runs` and need to install the custom controller 
- In *stable*, users rely on `Runs` and need to install the custom controller 
- In *integrated*, users don't rely on `Runs` and don't need the custom controller because the *Tekton Pipelines* controller will handle the execution

This is how a user specifies `Pipelines` in `Pipelines` :

```yaml
tasks:
...
- name: approve-slack-build-deploy
  runAfter:
    - integration-tests
  when:
    - input: $(params.git-action)
      operator: in
      values:
        - merge
  taskRef:
    apiVersion: tekton.dev/v1beta1
    name: approve-slack-build-deploy
    kind: Pipeline
```

In graduating to *integrated*, we can explore using the same syntax as is used in *stable* (as @mattmoor suggested in the API WG discussion)

Then, we can simplify the syntax, such as:

```yaml
tasks:
...
- name: approve-slack-build-deploy
  runAfter:
    - integration-tests
  when:
    - input: $(params.git-action)
      operator: in
      values:
        - merge
  pipelineRef:
    name: approve-slack-build-deploy
```

> 2. Do integrated custom tasks continue to rely on the `Run` CRD? If so, is the intention of the API changes to make that transparent to users?
> 

No, it no longer relies on the `Run` CRD -- it relies on the `Task`, `Pipeline`, `TaskRun` or `PipelineRun` CRDs depending on the functionality, such as, if its an authoring or runtime concern

> 3. In terms of code, does the custom task continue to live in a dedicated repo? Does the ownership move to the entire pipeline team? Is the custom task released together with pipeline ?
> 

In the graduation process from *stable* to *integrated*, we propose the `Custom Task` is deprecated in step 3.2 and removed from the dedicated repo in the last step:
https://github.com/tektoncd/community/blob/5dfaae796bf8d838fd86eddd62a15f09d38531a8/teps/0087-custom-tasks-graduation.md#L205-L222

> 4. What kind of backward compatibility we provide. If a custom task becomes integrated, do users have to change their pipelines in order to benefit from the latest versions?
> 

When a `Custom Task` is graduating from *stable* to *integrated*, we can require backwards compatibility in the functionality provided by the `Custom Task`, meaning:
- functionality in *stable* is exactly what will be provided when it is *integrated*
- if any changes need to be made based on user feedback, then it should be made in *stable* beforehand

Regarding the backwards compatibility in syntax, we can take the approach discussed in response to question 1 above

> 5. Finally, what level of stability is required for a custom task to have tasks and pipelines that depend on it in our catalog?

*stable* would be the level that we require for `Tasks` and `Pipelines` in the *Tekton Catalog* given its stability guarantees, what do you think? We can apply the same requirement for `Custom Tasks` support in the *Tekton Dashboard* (https://github.com/tektoncd/dashboard/issues/1970) and *Tekton CLI*

Reply:
> > This requirement is not 100% clear to me, I have a few questions.
> 
> Sure, happy to answer them and clarify in the TEP :)
> 
> > 1. Does this mean adding some syntactic sugar in the API so that it becomes easier for users to use?
> 
> It's more than that, it means that the functionality that the `Custom Task` provides is integrated into _Tekton Pipelines_, that is, the functionality is directly / natively supported in the _Tekton Pipelines API_ without the `Custom Task` / `Run`

I think what is important for an end user is the level of support they get for a certain feature. My list of requirements for integrated would be something like:
- when tekton is installed (via release file, operator) the feature is available with no extra step
- every new major / minor release of Tekton includes the feature - if an urgent fix is required for the feature, a new minor release of the entire Tekton pipeline is made
- the code is owned by the `tektoncd/pipeline` maintainer team

Other aspects like:
- the feature is managed by one of the existing controllers
- no extra CRD is needed

seems nice to have but I'm not convinced they should be hard requirements for the feature to be considered integrated.

Besides, for `Tasks` things like `Pipeline` and `Loop` ones, having `Run` seems like a nice thing, since the `Run` is the one resource where we can see params / results associated with "group" of `Tasks` and track the overall condition for the "group". 
We could do that in the original `PipelineRun` too, but what if we had a `Pipeline` that runs a `Pipeline` that runs a `TaskLoop` that runs a `Pipeline`? It can get too large to handle in the status of the original `PIpelineRun`.

> 
> Taking the example of `Pipelines` in `Pipelines`:
> 
> * In _stable_, users rely on `Runs` and need to install the custom controller
> * In _stable_, users rely on `Runs` and need to install the custom controller
> * In _integrated_, users don't rely on `Runs` and don't need the custom controller because the _Tekton Pipelines_ controller will handle the execution
> 
> This is how a user specifies `Pipelines` in `Pipelines` :
> 
> ```yaml
> tasks:
> ...
> - name: approve-slack-build-deploy
>   runAfter:
>     - integration-tests
>   when:
>     - input: $(params.git-action)
>       operator: in
>       values:
>         - merge
>   taskRef:
>     apiVersion: tekton.dev/v1beta1
>     name: approve-slack-build-deploy
>     kind: Pipeline
> ```

I think `Pipeline` is a bit of a special case of custom task, because it does not require an extra CRD, since `Pipeline` is already known to Tekton. Perhaps it would be good to include some examples using `TaskLoop` as well.

> 
> In graduating to _integrated_, we can explore using the same syntax as is used in _stable_ (as @mattmoor suggested in the API WG discussion)
> 
> Then, we can simplify the syntax, such as:
> 
> ```yaml
> tasks:
> ...
> - name: approve-slack-build-deploy
>   runAfter:
>     - integration-tests
>   when:
>     - input: $(params.git-action)
>       operator: in
>       values:
>         - merge
>   pipelineRef:
>     name: approve-slack-build-deploy
> ```
> 
> > 1. Do integrated custom tasks continue to rely on the `Run` CRD? If so, is the intention of the API changes to make that transparent to users?
> 
> No, it no longer relies on the `Run` CRD -- it relies on the `Task`, `Pipeline`, `TaskRun` or `PipelineRun` CRDs depending on the functionality, such as, if its an authoring or runtime concern
> 
> > 1. In terms of code, does the custom task continue to live in a dedicated repo? Does the ownership move to the entire pipeline team? Is the custom task released together with pipeline ?
> 
> In the graduation process from _stable_ to _integrated_, we propose the `Custom Task` is deprecated in step 3.2 and removed from the dedicated repo in the last step:

Is the intention that an integrated `CustomTask` becomes part of the stable API right away? If not, perhaps we should deprecate the stable custom task only once the integrated one is at least beta?
That said, we continued using `Run` for integrated tasks - at least in the beginning - it would make the process smoother:
- there would be no API change in the beginning, so the feature is added an is available as long as "Using CustomTasks in Pipeline" is enabled (currently alpha, we should bring it to beta?)
- when we introduce API sugar on top, we can do that as an alpha API change first, and then progress it to beta / stable as usual 

> 
> https://github.com/tektoncd/community/blob/5dfaae796bf8d838fd86eddd62a15f09d38531a8/teps/0087-custom-tasks-graduation.md#L205-L222
> 
> > 1. What kind of backward compatibility we provide. If a custom task becomes integrated, do users have to change their pipelines in order to benefit from the latest versions?
> 
> When a `Custom Task` is graduating from _stable_ to _integrated_, we can require backwards compatibility in the functionality provided by the `Custom Task`, meaning:
> 
> * functionality in _stable_ is exactly what will be provided when it is _integrated_
> * if any changes need to be made based on user feedback, then it should be made in _stable_ beforehand
> 
> Regarding the backwards compatibility in syntax, we can take the approach discussed in response to question 1 above
> 
> > 1. Finally, what level of stability is required for a custom task to have tasks and pipelines that depend on it in our catalog?
> 
> _stable_ would be the level that we require for `Tasks` and `Pipelines` in the _Tekton Catalog_ given its stability guarantees, what do you think? We can apply the same requirement for `Custom Tasks` support in the _Tekton Dashboard_ ([tektoncd/dashboard#1970](https://github.com/tektoncd/dashboard/issues/1970)) and _Tekton CLI_